data_exploration.ipynb

```diff
@@ -0,0 +1,1434 @@
z97Vse9s7e9Nx4ByKla6Ola0L3sUqUK/lz9y/cZrMTFlF0dbRw8TMWh4kEk2e3rtC+WrNvug3FETEkbM8Hj2bj3uOf7U0VfE9bFu8VCU69R9LsYAgrO2cCChXjdqNO3Dzykm1emZlZXHh8FqqN+pF0dI1sXfxpWXPySTEhnP/uvo8OndoDWWrtaBMcDNsHb1o0nkMOrp6/HV2p5Kctq4exmbW8qCnb5QvLT0DYyUZbR3dPHnzfdpKgNfPH3Ji3xra9R5XoNyRPRupWrsJVWo2wtHZg469h6Gjq8fZE3tVyh/bt5kSpYKo37Q9Ds7uNG/bG1cPP44f3CaX2b5hIf6lKvJTpwG4evhiY+9EYLmqmJhZyGWePgolpP5PePgUw8bOiUYtu2JgaMyLsAf/ER3NzK2Uwo0rZ/ErXhobOycA/EtVpNuA0RQPrICNnUz/uk3acf3yqQLzMzcn962jUkgzgmrI7N+6xx/o6OpxUZ39D26gaEBFajXuhL2TBw1b98PZvQins+uSvqExA0YtoXTFOtg6uuHu40/LbsN49ew+0RHvC63X58jKyuLq8bVU+aEXvoE1sXX2pXGXKSTEhvPwpvq65VUimOpNf8GvVK2vosOFI2upll2/7Vx8aZFdvx/cUK/DhcNrKFOtBaWDm2Hj6EXjTmPQ1tXjeq7+++DGyQTVakfVht2xdfLG2t6dEuXroZVrLBHSrD+V6nbCzln1KlKAo3s2EJyrXHbILpfnPlMu6zXtgIOzO82yy+WJg1vlMjs2LMC/VEVadvoZVw8/lXUHQEdXD1NzK3nQN8jfPsHXL4OQqw3yL68yjbCHt4iKeEeHfuNxdPXG0dWbjv3G8yrsPo/vql4RlJetew7wQ+0a1A+phpuLE7/27oaerg4Hj59WKT/y1/40rV8bbw83XJ0c+a1fT6TSLK7fviuXWb5+C+VLB9C7U1t8PNxxtLejUvkymJuZFkqn/1doSL5f+C/kv/NXCQRfiejoaA4fPkzfvn0xNDTMd9/MzAyAEiVKULZsWVauXKl0f9WqVVSsWBE/P9mSyA0bNhASEkJgYGC+tLS1tVU+4+9ibgTGBhqEvVOsLklNh7cRWThbF96DuF72+CFZzUcgewsN7C0l3Hii/ktDRno6r589wM9fMcCXSCT4+Zfn2ePbKuM8f3wb3zydcZGSFXmuRl4VCXFRvHhyB2NTC6aPaM/QbtWYNaozTx/cUBsnOuINCbGReBcLkl/TNzDGxdOfl09uqf59GWm8fX4f7+LKv8+7eJDaOGkpSfx1ZhcW1k6YWcqW81rbu2NgZMbV0zvIyEgjPS2Fq2d2YOPggZ2trcp00tPTefL0KaUCSio9OzAggPsP83/dVUVqaioZmZkYG6secAF8+pSEhoYGhkbqZQoiIz2dV2EPKJLLphKJhCL+5Xn2KFRlnLDHoUryAMUCg5TkpVIpK+f8QZ0mHXFw8fxivaLDs+2dy3YKe6suazn29imuKCMye1dQGwcgJTkBPX0jtRNBf53bi7auPv7la6u8/2/l32Tb5KREDI1M1N6PiXhDQlwkXrlsp2dgjLOHP6+eqrf3uxf38CqmbG/PYkG8enpLSfb2xf2M7x3E7KENObxlJmmpyfnS27t2PON7B7FgdEv+OrODrCxFG/292kqQfWlePWcoLbuNwNRc/TL4jPR0XoQ9pKh/OSUdi5UsR9ijOyrjPH10h6L+ylttSgRWkMtLpVJC/7qAnYML08f0p3/H2owb0onrl08rxfHy9efqhWMkJsQhlUq5fO4o6Wmp+BUv/c11zEtcbBSh188THNJY5f0cZGWycC9kGenpvHr2AN+89i9Rgedq6tLzx6FK5QWgaEBFnj9WLQ+QkpSIhoYG+obGhdKrMMRGviExLgL3IhXl1/QMjHH08Odt2K2v9pyCiIl4Q2JcJJ7FlOu309+o315FFfU7MT6K12GhGJlYsmRcayb2q8yyCe158ej6F+mXUy6L5Wkri5Ysx1N1beWjUKVyDFA8MEhF3XFl+ph+DOhYi/FDOnIjT90BuHT2EP3b1+SPAS3Ztm4+qakpKnX8T5TBfM/NSEMDDaVJKy0dXTQ0JDx9cPOz8dPTM3gc9pzSJRVbmiQSCaVLluDeo8eF0kE2FsrAxFg2FpdKpVz66ybODvYMHj2Rxh160GvwCM5dvvaZlAQCMckiEMjZv38/RkZG8tCiRQuePn1KVlaWfJKkILp27cq2bdtITEwEICEhge3bt9OlSxe5zJMnTwqVVl5SU1OJj49XCgUtvwUw0pdNpCSmKG/hSUzJwki/cM/VAOqV1eLlRynhsaq3ApX2lhAem8XrCPVbhRITYpBKMzE2Vd5Xa2xqSXxspMo48bGRmOSRNzFTL6+KyI+yfbYHty6iUkhz+o5YhLNHEeaN6074+5cq4yRkp29sqvySYWRqKb+Xl08JsUilmRjljWNiSUKccpyLxzYxoktpRnQtw8Pb5+g+bDlaWrJBhZ6+Ib3/WMONC/sY3qkUI7qU4dHt83T7bQmampoqnx0fH49UKsXMzFzpurmZGTExMSrj5GXFqtVYWlhQKiBA5f20tDRWrFpNtarBGBoYFCrNvOSUgbxf1ozNLImLVb0EOT42Mp+8iamy/JFdq5BoalKjQeu/pVeOfVTaO+5z9i58ef4UH8OxXYupUKOFWl2unt5BYMX68v39/1/4t9g2/P0rTh7cTHDtH9XK5NThvLYzMrUiIS5CZZwkdfY2UW4TAoJ+oGWvqXQfvoaqDXtw88Jeti7+TSlOSPP+tO43i66/r6B42VrsWTOOi0fXy+9/r7YSYPvqaXj4lqRk2eoFyiVk54dpPvtZEBej2t5xsVGY5vGrkFs+Pi6alJQkDuxcQ4lSQQwePY9SFaoxf8pvPLyreJHtM2QSmRkZ9GsfQvcWFVmzaCIDhk7D1t75m+uYlwsnD6Cnb0jpIPX59fH9a44f2EK1Ok3VyuRGXpfy2r8Ae8bHRmJsVvjykp6Wyq71syl
```

Comment:
This modifies our original `DataFrame`. We probably want to create a _view_ of the `DataFrame` and do the label encoding on that since we may want to use other encoding schemes like one-hot encoding for the actual training later.

We probably should enable copy-on write for Pandas as well (https://pandas.pydata.org/docs/dev/user_guide/copy_on_write.html).

Introduce a `ob_df_view = ob_df[:]`:

```suggestion
    "ob_df_view = ob_df[:]\n",
    "for column in ob_df_view.select_dtypes(include=['object']).columns:\n",
    "    label_encoders[column] = LabelEncoder()\n",
    "    ob_df_view[column] = label_encoders[column].fit_transform(ob_df[column])\n",
    "    \n",
    "# Calculate the correlation matrix\n",
    "correlation_matrix = ob_df_view.corr()\n",
```